File: src/services/abi.ts
/**
 * Minimal ABI encoding/decoding for EVM contract calls.
 * Handles Multicall3, ERC-20, Uniswap V3 QuoterV2, Solidly Router, and Chainlink feeds.
 * No external dependencies -- pure hex manipulation.
 */

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Left-pad a bigint/number to 32 bytes (64 hex chars). */
function h(v: bigint | number, _bytes = 32): string {
  return BigInt(v).toString(16).padStart(64, '0');
}

/** Left-pad an address to 32 bytes. */
function addr(a: string): string {
  return (a.startsWith('0x') ? a.slice(2) : a).toLowerCase().padStart(64, '0');
}

/** Right-pad hex data to a 32-byte boundary. */
function padR(hex: string): string {
  const c = hex.startsWith('0x') ? hex.slice(2) : hex;
  return c.padEnd(Math.ceil(c.length / 64) * 64, '0');
}

/** Read a 256-bit word from a hex string at a given hex-char position. */
function n64(hex: string, pos: number): number {
  const s = hex.slice(pos, pos + 64);
  if (!s) return 0;
  return Number(BigInt('0x' + s.padStart(64, '0')));
}

// ---------------------------------------------------------------------------
// Function selectors (first 4 bytes of keccak256 of canonical signature)
// ---------------------------------------------------------------------------

export const SEL = {
  aggregate3:       '82ad56cb', // aggregate3((address,bool,bytes)[])
  getEthBalance:    '4d2301cc', // getEthBalance(address)
  balanceOf:        '70a08231', // balanceOf(address)
  decimals:         '313ce567', // decimals()
  symbol:           '95d89b41', // symbol()
  name:             '06fdde03', // name()
  quoteV3:          'c6a5026a', // quoteExactInputSingle((address,address,uint256,uint24,uint160))
  getAmountsOut:    '5509a1ac', // getAmountsOut(uint256,(address,address,bool,address)[]) — Aerodrome/Velodrome V2 Router
  latestRoundData:  'feaf968c', // latestRoundData()
} as const;

// ---------------------------------------------------------------------------
// Multicall3 -- canonical deployment on all EVM chains
// ---------------------------------------------------------------------------

export const MULTICALL3 = '0xcA11bde05977b3631167028862bE2a173976CA11';

export interface Call3 {
  target: string;
  allowFailure: boolean;
  callData: string;
}

export interface McResult {
  success: boolean;
  data: string;
}

/**
 * ABI-encode an `aggregate3(Call3[])` call.
 *
 * Call3 = (address target, bool allowFailure, bytes callData).
 * Since the tuple contains `bytes` (dynamic), each element needs an offset pointer.
 */
export function encAggregate3(calls: Call3[]): string {
  // Function selector + offset to the single dynamic param (always 0x20)
  let out = SEL.aggregate3 + h(0x20);
  // Array length
  out += h(calls.length);

  // Pre-encode each element so we can compute byte-level offsets
  const elems = calls.map(c => {
    const cd = c.callData.startsWith('0x') ? c.callData.slice(2) : c.callData;
    const cdLen = cd.length / 2;
    return (
      addr(c.target) +              // address target    (32 bytes)
      h(c.allowFailure ? 1 : 0) +   // bool allowFailure (32 bytes)
      h(0x60) +                      // offset to bytes   (3*32 = 96)
      h(cdLen) +                     // bytes length
      padR(cd)                       // bytes data (right-padded)
    );
  });

  // Offset pointers (relative to start of elements area, which is right after the length word)
  let off = calls.length * 32; // skip past all offset slots
  for (const e of elems) {
    out += h(off);
    off += e.length / 2; // each hex pair = 1 byte
  }

  // Append element encodings
  out += elems.join('');

  return '0x' + out;
}

/**
 * ABI-decode the return value of `aggregate3 -> Result[]`.
 *
 * Result = (bool success, bytes returnData).
 */
export function decAggregate3(hex: string): McResult[] {
  const d = hex.startsWith('0x') ? hex.slice(2) : hex;
  if (d.length < 128) return [];

  const arrOff = n64(d, 0) * 2;          // byte offset -> hex-char offset
  if (arrOff + 64 > d.length) return [];

  const len = n64(d, arrOff);             // array length
  if (len <= 0 || len > 10_000) return []; // sanity cap
  const oStart = arrOff + 64;             // offsets begin after length word

  const out: McResult[] = [];
  for (let i = 0; i < len; i++) {
    const offPos = oStart + i * 64;
    if (offPos + 64 > d.length) break;

    const eOff = n64(d, offPos) * 2;
    const ePos = oStart + eOff;
    if (ePos + 128 > d.length) {
      out.push({ success: false, data: '0x' });
      continue;
    }

    const success = n64(d, ePos) !== 0;
    const bOff = n64(d, ePos + 64) * 2;   // offset to bytes within element
    const bPos = ePos + bOff;
    if (bPos + 64 > d.length) {
      out.push({ success, data: '0x' });
      continue;
    }

    const bLen = n64(d, bPos);             // bytes length
    const dataEnd = bPos + 64 + bLen * 2;
    const data = bLen > 0 && dataEnd <= d.length
      ? '0x' + d.slice(bPos + 64, dataEnd)
      : '0x';

    out.push({ success, data });
  }
  return out;
}

// ---------------------------------------------------------------------------
// ERC-20 call encoding
// ---------------------------------------------------------------------------

/** Encode `balanceOf(address)`. */
export function encBalanceOf(a: string): string {
  return '0x' + SEL.balanceOf + addr(a);
}

/** Encode `decimals()`. */
export function encDecimals(): string {
  return '0x' + SEL.decimals;
}

/** Encode Multicall3 `getEthBalance(address)`. */
export function encGetEthBalance(a: string): string {
  return '0x' + SEL.getEthBalance + addr(a);
}

// ---------------------------------------------------------------------------
// Uniswap V3 QuoterV2
// ---------------------------------------------------------------------------

/**
 * Encode `quoteExactInputSingle((address,address,uint256,uint24,uint160))`.
 * The struct is a static tuple, so encoding is just the concatenation of each field.
 */
export function encQuoteV3(
  tokenIn: string,
  tokenOut: string,
  amountIn: bigint,
  fee: number,
): string {
  return '0x' + SEL.quoteV3 +
    addr(tokenIn) +
    addr(tokenOut) +
    h(amountIn) +
    h(fee) +
    h(0n); // sqrtPriceLimitX96 = 0
}

// ---------------------------------------------------------------------------
// Aerodrome / Velodrome V2 Router
// ---------------------------------------------------------------------------

/**
 * Encode `getAmountsOut(uint256 amountIn, Route[] routes)` for a single-hop swap.
 * Route = (address from, address to, bool stable, address factory).
 * factory = 0x0 → use router's default factory.
 */
export function encGetAmountsOut(
  amountIn: bigint,
  tokenIn: string,
  tokenOut: string,
  stable: boolean,
): string {
  return '0x' + SEL.getAmountsOut +
    h(amountIn) +        // uint256 amountIn
    h(0x40) +            // offset to Route[] (2 head words × 32)
    h(1) +               // routes.length = 1
    addr(tokenIn) +      // route[0].from
    addr(tokenOut) +     // route[0].to
    h(stable ? 1 : 0) +  // route[0].stable
    h(0);                // route[0].factory = default (0x0)
}

/**
 * Decode the return of `getAmountsOut → uint256[]`.
 * For a single-hop route: amounts = [amountIn, amountOut].
 * Returns amountOut (the last element).
 */
export function decAmountsOut(hex: string): bigint {
  const d = hex.startsWith('0x') ? hex.slice(2) : hex;
  // Layout: offset(64) + length(64) + amounts[0](64) + amounts[1](64) = 256 hex chars
  if (d.length < 256) return 0n;
  return BigInt('0x' + d.slice(192, 256));
}

// ---------------------------------------------------------------------------
// Chainlink AggregatorV3
// ---------------------------------------------------------------------------

/** Encode `latestRoundData()` (no params). */
export function encLatestRound(): string {
  return '0x' + SEL.latestRoundData;
}

// ---------------------------------------------------------------------------
// Decoders
// ---------------------------------------------------------------------------

/** Decode the first 256-bit word as unsigned bigint. */
export function decU256(hex: string): bigint {
  const d = hex.startsWith('0x') ? hex.slice(2) : hex;
  if (d.length < 64) return 0n;
  return BigInt('0x' + d.slice(0, 64));
}

/** Decode the first 256-bit word as signed bigint. */
export function decI256(hex: string): bigint {
  const v = decU256(hex);
  return v >= (1n << 255n) ? v - (1n << 256n) : v;
}

/** Decode as uint8 (for `decimals()`). */
export function decU8(hex: string): number {
  return Number(decU256(hex));
}

/**
 * Decode an ABI-encoded `string` return value — the `[offset][length][data]`
 * layout that most ERC-20 `name()` / `symbol()` calls use. Falls back to a
 * `bytes32` interpretation for legacy tokens (e.g. MKR) that return a single
 * fixed 32-byte word, and decodes the bytes as UTF-8 so multibyte symbols
 * (e.g. "USD₮0") survive intact. Returns '' for short or empty input.
 */
export function decString(hex: string): string {
  const d = hex.startsWith('0x') ? hex.slice(2) : hex;
  if (d.length < 64) return '';
  // bytes32-style: a single 32-byte word, no offset/length header.
  if (d.length < 128) return utf8FromHex(d.slice(0, 64));
  const len = n64(d, 64);
  // A real dynamic string declares its byte length in the 2nd word; an
  // out-of-range value means this isn't offset-encoded → treat head as bytes32.
  if (len <= 0 || len > 4096 || 128 + len * 2 > d.length) return utf8FromHex(d.slice(0, 64));
  return utf8FromHex(d.slice(128, 128 + len * 2));
}

/** Decode a run of hex-encoded bytes as UTF-8, stopping at the first NUL. */
function utf8FromHex(dataHex: string): string {
  const bytes: number[] = [];
  for (let i = 0; i + 1 < dataHex.length; i += 2) {
    const b = parseInt(dataHex.slice(i, i + 2), 16);
    if (b === 0) break; // NUL = end of bytes32 padding / C-string terminator
    bytes.push(b);
  }
  let out = '';
  for (let i = 0; i < bytes.length; ) {
    const b0 = bytes[i];
    if (b0 < 0x80) { out += String.fromCharCode(b0); i += 1; }
    else if (b0 < 0xc0) { i += 1; } // stray continuation byte → skip
    else if (b0 < 0xe0 && i + 1 < bytes.length) {
      out += String.fromCharCode(((b0 & 0x1f) << 6) | (bytes[i + 1] & 0x3f));
      i += 2;
    } else if (b0 < 0xf0 && i + 2 < bytes.length) {
      out += String.fromCharCode(((b0 & 0x0f) << 12) | ((bytes[i + 1] & 0x3f) << 6) | (bytes[i + 2] & 0x3f));
      i += 3;
    } else if (i + 3 < bytes.length) {
      const cp = ((b0 & 0x07) << 18) | ((bytes[i + 1] & 0x3f) << 12) | ((bytes[i + 2] & 0x3f) << 6) | (bytes[i + 3] & 0x3f);
      const c = cp - 0x10000;
      out += String.fromCharCode(0xd800 + (c >> 10), 0xdc00 + (c & 0x3ff));
      i += 4;
    } else { i += 1; }
  }
  return out.trim();
}

/**
 * Decode Chainlink `latestRoundData()` return value.
 * Returns: (uint80 roundId, int256 answer, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound)
 * All Chainlink USD feeds use 8 decimals.
 */
export function decChainlinkUsd(hex: string): number {
  const d = hex.startsWith('0x') ? hex.slice(2) : hex;
  if (d.length < 128) return 0;
  // answer is the 2nd word (int256, but always positive for USD feeds)
  const answer = BigInt('0x' + d.slice(64, 128));
  return Number(answer) / 1e8;
}

/**
 * Decode the raw `answer` (2nd word, int256) from Chainlink `latestRoundData()`
 * WITHOUT scaling. Pair with the feed's `decimals()` to compute the price
 * (`answer / 10**decimals`) — fiat/USD feeds vary (most are 8, e.g. PHP is 18).
 */
export function decChainlinkAnswer(hex: string): bigint {
  const d = hex.startsWith('0x') ? hex.slice(2) : hex;
  if (d.length < 128) return 0n;
  const raw = BigInt('0x' + d.slice(64, 128));
  return raw >= 1n << 255n ? raw - (1n << 256n) : raw; // int256 sign
}

File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/wallet/core/Abi.kt
package app.getvela.wallet.feature.wallet.core

import java.math.BigInteger

/**
 * Minimal ABI encoding and decoding for the read path: Multicall3, ERC-20,
 * Uniswap-V3 QuoterV2, Solidly routers and Chainlink feeds. Pure hex, no
 * dependency.
 *
 * Port of `app-web/vela-wallet/src/lib/services/abi.ts`.
 *
 * **Why this is a shell file and not a core one.** The codebase already drew
 * this line and it is worth stating rather than re-deciding: the *judgement*
 * about a price — which pool wins inside a quote token, which stable wins
 * across them, whether a DEX quote survives the sanity band — lives in
 * `vela_core::app::balance_dashboard` and is called from here through uniffi.
 * What is in this file is transport framing: 32-byte words and offsets. The
 * web made the same split deliberately.
 *
 * **What that costs, honestly.** This is a second hand-written copy of the
 * framing, and there is no generated artifact to drift-test it against — the
 * way `CoreWireDriftTest` guards the wire types. The mitigation is that every
 * function here is pinned against a known encoding in `AbiTest`, and the whole
 * path is exercised against real chains on a device. `first_grouped_quote_price`
 * — the custom-token rule, which IS judgement and today exists only in the
 * web's TypeScript — is deliberately NOT ported here; it belongs in the core
 * before Android prices a custom token.
 */
object Abi {

    /** The canonical Multicall3 deployment, identical on every EVM chain. */
    const val MULTICALL3 = "0xcA11bde05977b3631167028862bE2a173976CA11"

    // First 4 bytes of keccak256 over the canonical signature.
    private const val SEL_AGGREGATE3 = "82ad56cb" // aggregate3((address,bool,bytes)[])
    private const val SEL_GET_ETH_BALANCE = "4d2301cc" // getEthBalance(address)
    private const val SEL_BALANCE_OF = "70a08231" // balanceOf(address)
    private const val SEL_DECIMALS = "313ce567" // decimals()
    private const val SEL_QUOTE_V3 = "c6a5026a" // quoteExactInputSingle((address,address,uint256,uint24,uint160))
    private const val SEL_GET_AMOUNTS_OUT = "5509a1ac" // getAmountsOut(uint256,(address,address,bool,address)[])
    private const val SEL_SYMBOL = "95d89b41" // symbol()
    private const val SEL_NAME = "06fdde03" // name()
    private const val SEL_ALLOWANCE = "dd62ed3e" // allowance(address,address)
    private const val SEL_LATEST_ROUND = "feaf968c" // latestRoundData()

    /** One entry in a Multicall3 batch. `allowFailure` is always true here. */
    data class Call(val target: String, val callData: String)

    /** One answer from a batch: whether the call succeeded, and its return data. */
    data class CallResult(val success: Boolean, val data: String)

    // -- encoding ------------------------------------------------------------

    /**
     * `aggregate3(Call3[])`, where `Call3 = (address, bool, bytes)`.
     *
     * The tuple holds `bytes`, which is dynamic, so each element is encoded
     * first and then addressed by a byte offset measured from the start of the
     * element area — the word right after the array length.
     */
    fun encodeAggregate3(calls: List<Call>): String {
        val elements = calls.map { call ->
            val data = strip(call.callData)
            word(BigInteger(strip(call.target).lowercase(), 16)) + // address target
                word(BigInteger.ONE) + //                            bool allowFailure
                word(BigInteger.valueOf(0x60)) + //                   offset to bytes (3 words)
                word(BigInteger.valueOf((data.length / 2).toLong())) + // bytes length
                padRight(data)
        }

        val head = StringBuilder()
        head.append(SEL_AGGREGATE3)
        head.append(word(BigInteger.valueOf(0x20))) // offset to the one dynamic param
        head.append(word(BigInteger.valueOf(calls.size.toLong())))

        var offset = calls.size * 32L // past the offset slots themselves
        elements.forEach { element ->
            head.append(word(BigInteger.valueOf(offset)))
            offset += element.length / 2
        }
        elements.forEach(head::append)
        return "0x$head"
    }

    fun encodeBalanceOf(address: String): String = "0x" + SEL_BALANCE_OF + addressWord(address)

    fun encodeDecimals(): String = "0x$SEL_DECIMALS"

    fun encodeSymbol(): String = "0x$SEL_SYMBOL"

    fun encodeName(): String = "0x$SEL_NAME"

    fun encodeAllowance(owner: String, spender: String): String = "0x" + SEL_ALLOWANCE + addressWord(owner) + addressWord(spender)

    fun encodeGetEthBalance(address: String): String =
        "0x" + SEL_GET_ETH_BALANCE + addressWord(address)

    fun encodeLatestRound(): String = "0x$SEL_LATEST_ROUND"

    /**
     * Uniswap-V3 `quoteExactInputSingle`. The struct is static, so the encoding
     * is the fields in order; `sqrtPriceLimitX96` is 0 (no limit).
     */
    fun encodeQuoteV3(
        tokenIn: String,
        tokenOut: String,
        amountIn: BigInteger,
        fee: Int,
    ): String = "0x" + SEL_QUOTE_V3 +
        addressWord(tokenIn) +
        addressWord(tokenOut) +
        word(amountIn) +
        word(BigInteger.valueOf(fee.toLong())) +
        word(BigInteger.ZERO)

    /**
     * Solidly (Aerodrome/Velodrome V2) `getAmountsOut(uint256, Route[])` for a
     * single hop. `Route = (address from, address to, bool stable, address
     * factory)`; a zero factory means the router's default.
     */
    fun encodeGetAmountsOut(
        amountIn: BigInteger,
        tokenIn: String,
        tokenOut: String,
        stable: Boolean,
    ): String = "0x" + SEL_GET_AMOUNTS_OUT +
        word(amountIn) +
        word(BigInteger.valueOf(0x40)) + // offset to Route[] — two head words
        word(BigInteger.ONE) + //           routes.length
        addressWord(tokenIn) +
        addressWord(tokenOut) +
        word(if (stable) BigInteger.ONE else BigInteger.ZERO) +
        word(BigInteger.ZERO)

    // -- decoding ------------------------------------------------------------

    /**
     * `aggregate3 → Result[]`, where `Result = (bool success, bytes data)`.
     *
     * Every bound is checked because this parses a remote answer: a truncated
     * or hostile response must produce fewer results, never an exception and
     * never a wrong number. A missing element is reported as a failed call,
     * which is what the core already knows how to handle.
     */
    fun decodeAggregate3(hex: String): List<CallResult> {
        val d = strip(hex)
        if (d.length < 128) return emptyList()

        val arrayOffset = wordAt(d, 0) * 2
        if (arrayOffset < 0 || arrayOffset + 64 > d.length) return emptyList()

        val count = wordAt(d, arrayOffset)
        if (count <= 0 || count > 10_000) return emptyList() // sanity cap
        val offsetsStart = arrayOffset + 64

        val results = ArrayList<CallResult>(count)
        for (index in 0 until count) {
            val offsetPos = offsetsStart + index * 64
            if (offsetPos + 64 > d.length) break

            val elementPos = offsetsStart + wordAt(d, offsetPos) * 2
            if (elementPos < 0 || elementPos + 128 > d.length) {
                results.add(CallResult(success = false, data = "0x"))
                continue
            }

            val success = wordAt(d, elementPos) != 0
            val bytesPos = elementPos + wordAt(d, elementPos + 64) * 2
            if (bytesPos < 0 || bytesPos + 64 > d.length) {
                results.add(CallResult(success, "0x"))
                continue
            }

            val length = wordAt(d, bytesPos)
            val end = bytesPos + 64 + length * 2
            val data = if (length > 0 && end <= d.length && end > bytesPos + 64) {
                "0x" + d.substring(bytesPos + 64, end)
            } else {
                "0x"
            }
            results.add(CallResult(success, data))
        }
        return results
    }

    /** The first 256-bit word, unsigned. Short input reads as zero, not as an error. */
    fun decodeUint256(hex: String): BigInteger {
        val d = strip(hex)
        if (d.length < 64) return BigInteger.ZERO
        return runCatching { BigInteger(d.substring(0, 64), 16) }.getOrDefault(BigInteger.ZERO)
    }

    /** `decimals()`. */
    fun decodeUint8(hex: String): Int = decodeUint256(hex).toInt()

    /**
     * An ABI `string` return — `symbol()` and `name()`.
     *
     * Two layouts in the wild: the modern `[offset][length][data]`, and a
     * single fixed 32-byte word from legacy tokens that declared `bytes32`
     * (MKR is the famous one). A length that cannot be a length means this is
     * the legacy shape, not a corrupt string.
     *
     * Decoded as UTF-8 so multibyte symbols survive — "USD₮0" is a real
     * token, and rendering it as mojibake makes a legitimate holding look like
     * a scam.
     *
     * Returns `null` rather than `""` when nothing readable is there: an
     * unresolvable symbol is a fact the core acts on, and an empty string
     * would read as a token that answered with a blank name.
     */
    fun decodeString(hex: String): String? {
        val d = strip(hex)
        if (d.length < 64) return null
        if (d.length < 128) return utf8(d.substring(0, 64))

        val length = wordAt(d, 64)
        if (length <= 0 || length > 4096 || 128 + length * 2 > d.length) {
            return utf8(d.substring(0, 64))
        }
        return utf8(d.substring(128, 128 + length * 2))
    }

    /** Hex bytes as UTF-8, stopping at the first NUL (bytes32 padding). */
    private fun utf8(dataHex: String): String? {
        val bytes = ArrayList<Byte>(dataHex.length / 2)
        var index = 0
        while (index + 1 < dataHex.length) {
            val byte = dataHex.substring(index, index + 2).toIntOrNull(16) ?: break
            if (byte == 0) break
            bytes.add(byte.toByte())
            index += 2
        }
        if (bytes.isEmpty()) return null
        return String(bytes.toByteArray(), Charsets.UTF_8).trim().ifBlank { null }
    }

    /**
     * `getAmountsOut → uint256[]`. For one hop the array is
     * `[amountIn, amountOut]`, and the answer is the last element:
     * `offset | length | amounts[0] | amounts[1]`.
     */
    fun decodeAmountsOut(hex: String): BigInteger {
        val d = strip(hex)
        if (d.length < 256) return BigInteger.ZERO
        return runCatching { BigInteger(d.substring(192, 256), 16) }
            .getOrDefault(BigInteger.ZERO)
    }

    /**
     * Chainlink `latestRoundData → (roundId, answer, startedAt, updatedAt,
     * answeredInRound)` — the USD price is `answer`, the second word, with the
     * feeds' fixed 8 decimals.
     *
     * `answer` is `int256`: a negative value is a feed reporting something this
     * cannot interpret as a price, and it must not wrap into an enormous
     * positive one.
     *
     * **A deliberate divergence from the web port.** `decChainlinkUsd` there
     * reads the word as unsigned ("always positive for USD feeds") while its
     * sibling `decChainlinkAnswer` applies the sign. A negative answer would
     * therefore decode as ~1.16e69 and sail through the caller's `> 0` gate.
     * The feeds are positive in practice, so this is defence rather than a
     * fixed bug — but a wallet should not have a path where one bad word prices
     * a coin at 10^69 dollars.
     *
     * Returns `null` rather than a number for anything unusable, which is the
     * `isFinite && > 0` gate `choose_native_price` documents as the shell's.
     */
    fun decodeChainlinkUsd(hex: String): Double? = decodeChainlinkAnswer(hex, USD_FEED_DECIMALS)

    /**
     * The same answer, scaled by a decimals the caller read from the feed.
     *
     * The USD price feeds all use 8, but the FIAT feeds do not — PHP reports 18
     * — so anything reading a currency feed must pass what that feed said about
     * itself. Assuming 8 there is a rate off by a factor of 10^10.
     */
    fun decodeChainlinkAnswer(hex: String, decimals: Int): Double? {
        val d = strip(hex)
        if (d.length < 128) return null
        if (decimals < 0 || decimals > 38) return null
        val raw = runCatching { BigInteger(d.substring(64, 128), 16) }.getOrNull() ?: return null
        val signed = if (raw.bitLength() >= 256) raw.subtract(TWO_POW_256) else raw
        if (signed.signum() <= 0) return null
        val price = signed.toDouble() / StrictMath.pow(10.0, decimals.toDouble())
        return if (price.isFinite() && price > 0.0) price else null
    }

    // -- helpers -------------------------------------------------------------

    private val TWO_POW_256: BigInteger = BigInteger.ONE.shiftLeft(256)

    /** Every Chainlink USD PRICE feed reports 8. The fiat feeds do not. */
    private const val USD_FEED_DECIMALS = 8

    private fun strip(hex: String): String = if (hex.startsWith("0x")) hex.substring(2) else hex

    /** A 32-byte left-padded word. */
    private fun word(value: BigInteger): String = value.toString(16).padStart(64, '0')

    private fun addressWord(address: String): String =
        strip(address).lowercase().padStart(64, '0')

    /** Right-pad dynamic data to a 32-byte boundary. */
    private fun padRight(hex: String): String {
        val words = (hex.length + 63) / 64
        return hex.padEnd(words * 64, '0')
    }

    /**
     * A 256-bit word read as an Int, for lengths and offsets.
     *
     * These are positions inside a byte string this device just received, so a
     * value that cannot be a position — negative, or beyond any plausible
     * response — answers -1 and every caller bounds-checks it. Truncating a
     * huge word into a valid-looking index is how a decoder reads a response as
     * something it is not.
     */
    private fun wordAt(hex: String, position: Int): Int {
        if (position < 0 || position + 64 > hex.length) return -1
        val value = runCatching { BigInteger(hex.substring(position, position + 64), 16) }
            .getOrNull() ?: return -1
        if (value.signum() < 0 || value.bitLength() > 31) return -1
        return value.toInt()
    }
}
